### DISCUSSION TITLE
The Miser Journey's Heart: Taming Representation and Layering of Abstractions

### DISCUSSION THREAD

**Discussion by orcmid** [orcmid/miser#58] [General]
I am struggling to capture how representation is accomplished and how we appreciate computing in the world while overcoming the tendency to magical thinkng and caegory errors.

This all started for me in early work on compilers and programming language work observed by me in the early 1960s around the time that OS/360 was being introduced and particular ways of treating representations in data via the OS and how input-output interfaces were deliverd in code.  There were essentially two cases: partitioned datasets, which were used to house code libraries, and the record-oriented datasets that were for general input-output, with random-access maybe not so much until HDDs became more prominent.  Making continuous text streams a key data structure came along with C and Unix although compilers operated in that manner with regard to source-code text and also some binary structures (e.g., Zip packages now).

Treatment of removability of quasi-persistent data sets (as streamed from HDDs or magnetic tapes) also involved metadata, whether incorporated in the data or provided separately.  This led to "meta-" issues and something similar arose in the OSI 7-level stack, where administration aspects were cross-cutting and not their own layer.  Meta-object and aspect-oriented programming were pretty late coming to this party.

My original thinking was not about representations of natural entities, something that databases, programming-language objects, and mechanical deduction systems rub up against.  I was just concerned on how the levels of abstraction still all about the (reductive) computational schemes could be apprehended.

I have also been disheartened about how we confuse data as being about something when it is so syntactically constrained and devoid of inherent interpretation.  Intentionality is elsewhere.  Something like William Kent's "Data and Reality" is more toward what we need to understand.

There is great success with systems that allow symbolic associations without addressing semantics very well.  In some sense, this applies to database models, starting with Codd's Relational Model, since it is purely about data.  The same applies to SQL and also arrangements such as the way labelled data arises in various schemes, including a variety of Microsoft APIs where strings and name value-pair arrangements matter.  What is amazing to me is the amount that is engineered, and good enough, with poor syntactics and not much awareness of semantics and ontological commitment (including reality agreement) that goes on.

We will probably not stop doing that.  I claim it would be valuable to be doing that on purpose and in a ways where the intention to serve human purposes is made accountable and dependable.  

Pondering.

**Comment by band** [comment]
This really is amazing: "What is amazing to me is the amount that is engineered, and good enough, with poor syntactics and not much awareness of semantics and ontological commitment (including reality agreement) that goes on."

Maybe reality agreement seems poor because the structural needs of the technology are a poor fit. But for some amazing reasons what we are doing is good enough to provide amazing technology and associated affordances. 🤔

**Comment by rnd0101** [comment]
It's probably futile to make this absolute. Every model (layers of abstractions) is supposed to be relative to the purpose. I imagine there is some law, similar to Heisenberg's uncertainty principle in physics, which sets a limit for semantics to be universal enough and sharp at the same time for some useful pragmatics. This is also similar to what is being talked in the Brandom's "Between Saying & Doing" book I mentioned previously, that only interlacing "saying" and "doing" can bring useful abstractions. In other words, it's impossible to establish "absolute" semantics by using some symbolic system, even layering abstractions. In reality, all abstractions leak. Sometimes, badly.

Probably, some balance is needed on the **meaning-use** axis. Meaning and use interact. They define each other. Wave and particle at the same time, depending on the observer.

I feel, that this is something very **fundamental**, and the only practical way to deal with it is necessary relative (and stopping at true enough approximations), or then using some dialectical reasoning until good approximation is reached for the model.

One interesting example is Category Theory. This one is built totally on assuming relativity, while still being very useful. Instead of going through the pile of turtles (manifested in well-known set theory paradoxes), CT embraced the uncertainty to produce quite universal (math-wide) results, which also attached highly abstract meanings to things, which were previously known under different names in multiple branches of mathematics.

There most likely are no universal meanings at all. Best we can do is to make some bridges. This is especially true for "the intention to serve human purposes is made accountable and dependable".
